### DISCUSSION TITLE
Queries: is it possible to match any number of tree levels?

### DISCUSSION THREAD

**Discussion by przepompownia** [nvim-treesitter/nvim-treesitter-textobjects#336] [Q&A]
Inspiration: how to match a call expression that consist of the chain of methods, but with the cursor at the start of such expression?

Example from PHP
```php
        User::find()
            ->joinWith('xxx')
            ->someOtherMethod()
            ->all();
```
and its part of the tree that needs matching
```
     (member_call_expression) [19:9-22:19]
      object: (member_call_expression) [19:9-21:31]
       object: (member_call_expression) [19:9-20:29]
        object: (scoped_call_expression) [19:9-20]
```

I have read the basics from https://tree-sitter.github.io/tree-sitter/using-parsers#query-syntax but not found any syntax to express it. I managed to match only concrete number of levels with e.g.
```
(member_call_expression 
  ([
	(scoped_call_expression)
	(member_call_expression)
	])?
) @call.outer
```

**Comment by theHamsta** [comment]
No. That's not possible at the moment. You could use `#has-ancestor?` which can walk up arbitrarily many objects up in the hierarchy, but at some performance cost.

**Reply by przepompownia** [reply]
@theHamsta thanks to your hint I looked around where `#has-ancestor?` and other predicates can be defined at the Neovim and its extensions level and also found the help text under `treesitter-predicates`. 

After many tries and errors, I ended up with
```
[
 (member_call_expression)
 (scoped_call_expression)
] @_class_call_expression

(
  (member_call_expression) @_mc
  (#has-ancestor? @_class_call_expression @_mc)
) @call.outer
```
but it seems to stop at the shortest chain i.e like in the two-level example above.
